src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -465,51 +487,62 @@ private static unsafe OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, o
             InvalidExit:
                 bytesWritten = (int)destIndex;
                 return ignoreWhiteSpace ?
-                    DecodeWithWhiteSpaceFromUtf8InPlace(buffer, ref bytesWritten, sourceIndex) : // The input may have whitespace, attempt to decode while ignoring whitespace.
+                    DecodeWithWhiteSpaceFromUtf8InPlace<TBase64Decoder>(buffer, ref bytesWritten, sourceIndex) : // The input may have whitespace, attempt to decode while ignoring whitespace.
                     OperationStatus.InvalidData;
             }
         }
 
-        private static OperationStatus DecodeWithWhiteSpaceBlockwise(ReadOnlySpan<byte> utf8, Span<byte> bytes, ref int bytesConsumed, ref int bytesWritten, bool isFinalBlock = true)
+        internal static OperationStatus DecodeWithWhiteSpaceBlockwise<TBase64Decoder>(ReadOnlySpan<byte> source, Span<byte> bytes, ref int bytesConsumed, ref int bytesWritten, bool isFinalBlock = true)
+            where TBase64Decoder : IBase64Decoder<byte>
         {
             const int BlockSize = 4;
             Span<byte> buffer = stackalloc byte[BlockSize];
             OperationStatus status = OperationStatus.Done;
 
-            while (!utf8.IsEmpty)
+            while (!source.IsEmpty)
             {
                 int encodedIdx = 0;
                 int bufferIdx = 0;
                 int skipped = 0;
 
-                for (; encodedIdx < utf8.Length && (uint)bufferIdx < (uint)buffer.Length; ++encodedIdx)
+                for (; encodedIdx < source.Length && (uint)bufferIdx < (uint)buffer.Length; ++encodedIdx)
                 {
-                    if (IsWhiteSpace(utf8[encodedIdx]))
+                    if (IsWhiteSpace(source[encodedIdx]))
                     {
                         skipped++;
                     }
                     else
                     {
-                        buffer[bufferIdx] = utf8[encodedIdx];
+                        buffer[bufferIdx] = source[encodedIdx];
                         bufferIdx++;
                     }
                 }
 
-                utf8 = utf8.Slice(encodedIdx);
+                source = source.Slice(encodedIdx);
                 bytesConsumed += skipped;
 
                 if (bufferIdx == 0)
                 {
                     continue;
                 }
 
-                bool hasAnotherBlock = utf8.Length >= BlockSize && bufferIdx == BlockSize;
+                bool hasAnotherBlock;
+
+                if (typeof(TBase64Decoder) == typeof(Base64DecoderByte) || bufferIdx == 1)
+                {
+                    hasAnotherBlock = source.Length >= BlockSize && bufferIdx == BlockSize;
+                }
+                else
+                {
+                    hasAnotherBlock = source.Length > 1;
+                }
```

Comment:
This condition is odd.

Doesn't `source` not being empty imply that `bufferIdx == BufferSize`?
Therefore `source.Length >= BlockSize && bufferIdx == BlockSize` could be simplified to `source.Length >= BlockSize`.

The `|| bufferIdx == 1` part would also imply that `hasAnotherBlock` should be `false`, which could also be achieved by the `source.Length > 1` part.

Reply:
I think you are right for both